keps/sig-apps/4026-job-creation-timestamp-annotation/README.md

```diff
@@ -0,0 +1,299 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, there is no supported way to get the original/expected initial scheduled timestamp for the job created from a cronjob. This KEP proposes to set the the original scheduled time as an annotation in the job metadata.
+
+## Motivation
+
+### Goals
+
+- Set job scheduled timestamp as an annotation on the job.
+- Adding the annotation should not be disruptive to existing workloads.
+
+### Non-Goals
+
+## Proposal
+
+At a high level, the proposal is to modify the CronJob controller to set the job scheduled timestamp as a job annotation. The details of this are outlined in the Design Details section below.
+
+Job scheduled timestamp annotation: `batch.kubernetes.io/cronjob-scheduled-timestamp`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I would like to get the job's scheduled timestamp that this job was expected to be running.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+CronJobs are always working with the assumption that the changes apply only to newly created jobs after the change. Therefore, the change will be to inject the annotation for newly created Jobs from CronJobs for when the feature is on. This will nicely play with downgrade and doesn't introduce unnecessary complexity.
+
+## Design Details
+
+The CronJob controller will only need a minor update to the [getJobFromTemplate2](https://github.com/kubernetes/kubernetes/blob/7024beeeeb1f2e4cde93805a137cd7ad92fec466/pkg/controller/cronjob/utils.go#L188) function, to add the job scheduled timestamp as the job annotation `batch.kubernetes.io/cronjob-scheduled-timestamp`. The scheduled timestamp is represented in `RFC3339`.
+
+For the scheduled timestamp's timezone, the initial thought was to use `UTC` as it's used as the primary one for less confusion. However, since the `job` object has a `spec.timeZone`, it was a better to use the same timezone within the same object. If the job `spec.timeZone` is not set or `nil`, the annotation will use the `UTC` timezone as a default.
+
+### Test Plan
+
+- [X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+##### Unit tests
+
+- `k8s.io/kubernetes/pkg/controller/cronjob`: `05/22/2023` - `96.2%`
+
+##### Integration tests
+
+Unit tests will ensure the new annotation is correctly added to jobs, and integration tests will verify that the annotation is only added to jobs from newly created CronJobs, not existing workloads.
```

Comment:
for newly created Jobs for any CronJobs, right?

[i.e. we create it for jobs created from both newly added and already existing CronJobs, just don't touch already existing jobs, right?]

Reply:
I missed this, in the risks and mitigations section we explicitly pointed out that `inject the annotation for newly created Jobs from CronJobs for when the feature is on` so this test should ensure the annotation is present when the feature is on and missing when off.

Reply:
Updated.